Native notifications for Windows, Linux and OSX

pr8x opened the pull request:
## What does the pull request do?
Implementing native notification support for Windows (10), Linux (DBUS/FreeDesktop) and OSX.

![linux_native_notifications](https://user-images.githubusercontent.com/4670166/71370093-10d22980-25ad-11ea-8b41-424f52e78e0d.png)
___

![windows_native_notifications](https://user-images.githubusercontent.com/4670166/71370096-12035680-25ad-11ea-8215-9a65a65c9afc.png)

___

![osx_native_notification](https://user-images.githubusercontent.com/4670166/71428627-300eab00-26c2-11ea-802a-c75ec5d129a3.png)


## Discussion
- In Windows it is required to register an application shortcut in the StartMenu in order to send toast notifications (See https://docs.microsoft.com/en-us/windows/win32/shell/enable-desktop-toast-with-appusermodelid). I am not exactlly sure how we want to handle this. I could add a `RegisterApplication` function in the `Win32Platform` to do this registration and also create a unique `AppUserModelID` based on the `Application.Name`. It's probably a good idea to uninstall the shortcut once the application exits.

- This PR provides a *basic* implementation that works across the three major desktop platforms. Advanced features (such as replacing notifications, custom buttons, images, etc) can be rolled in future updates. 


## What is the current behavior?
We only have managed notifications hosted inside avalonia windows.


## What is the updated/expected behavior with this PR?
<!--- Describe how to test the PR. -->

## How was the solution implemented (if it's not obvious)?
- Windows 10: Using the UWP `ToastNotification` API
- Linux: Using the FreeDesktop DBUS `org.freedesktop.Notifications` service interface. (Thanks to #2884)
- OSX: Creating a COM interface for interacting with the native Cocoa API (`NSUserNotificationCenter`) 



## Fixed issues
#2155

pr8x committed:
!F Porting FreeDesktop NotificationManager from PR

pr8x committed:
!T Cleaning up the notification page

pr8x committed:
!B Fixing About and Exit commands

pr8x committed:
!F Adding support for Win10NotificationManager

pr8x committed:
!F OSX Notification COM implementation

kekekeks commented on src/Windows/Avalonia.Win32/Avalonia.Win32.csproj:
```diff
@@ -4,6 +4,9 @@
     <PackageId>Avalonia.Win32</PackageId>
   </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Windows.SDK.Contracts" Version="10.0.18362.2005" />
```
Does that work with Win 8.1 or is it Win10+ only?

jkoritzinsky replied:
This is Win10+ only from what I can tell.

pr8x replied:
> The Windows 10 WinRT API Pack enables you to add the latest Windows Runtime APIs support to your .NET Framework 4.5+ and .NET Core 3.0+ libraries and apps.

https://www.nuget.org/packages/Microsoft.Windows.SDK.Contracts/10.0.17763.1000-preview

But AFAIK the toast notifications only work in Windows 10 anyway.

MichalPetryka replied:
Windows 8.1 also has them

pr8x replied:
Yeah Windows 8 and 8.1 both have support for the ToastNotification API. Not sure if we really care about Windows 8 though 😄 .

kekekeks replied:
We probably want a managed fallback that displays notifications in separate windows manually. Like IM clients of the past did. That's not in the scope of this PR though.

MichalPetryka replied:
Well, it's more of "do people who use Avalonia for their software care about Windows 8 and 8.1", not "if we care about it".

pr8x committed:
!F Adding CreateNotificationManager hook to IAvaloniaNativeFactory

!T FIxed warning

pr8x committed:
!F Implementing NativeNotificationManager

!F Close and action callbacks
!F Handling notification center events in seperate delegate
!T Renaming some files to be more consistent

pr8x committed:
!T Reverting to .NET Core 2

pr8x commented:
@danwalmsley So how do we solve the problem with registering the application shortcut in the startmenu on Windows? Should I just register a temporary and uninstall it when the application terminates?

MonkAlex commented:
its not required, but will be used, if registered (not check, only google)
[social.msdn.microsoft.com](https://social.msdn.microsoft.com/Forums/sqlserver/en-US/5a6c5331-225f-4528-995a-99091f8070e0/raise-a-toast-notification-from-a-desktop-app-without-a-valid-shortcut-installed-in-the-start-screen?forum=windowsgeneraldevelopmentissues)

pr8x commented:
In my experience deleting the shortcut (even when the _AppUserModelID_ has been assigned to the current process via `SetCurrentProcessExplicitAppUserModelID`) will cause the toast nofications to be ignored by Windows. Most official Microsoft sources confirm this. 
> A shortcut to your app, with a System.AppUserModel.ID, must be installed to the Start screen. Note, however, that it does not need to be pinned to the Start screen

https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/hh802768(v=vs.85)?redirectedfrom=MSDN

Also it seems like once the shortcut was installed they use the name and the icon of it to display the application in the taskbar. Without a shortcut the taskbar icon will default to the main window icon AFAICT. So it might actually be a good idea to have some sort of `RegisterApplication` function on the `Win32Platform` that takes care of registering your _AppUserModelID_ and assigning name and icon of your application via the shortcut.

pr8x committed:
Merge branch 'master' into native_notifications_win_linux_osx

# Conflicts:
#	src/Avalonia.Native/AvaloniaNativePlatform.cs

grokys commented:
@danwalmsley this is ready for review now, when you have a moment.

Commit:
Merge branch 'master' into native_notifications_win_linux_osx

danwalmsley commented on samples/ControlCatalog/MainWindow.xaml:
```diff
@@ -45,10 +45,10 @@
     <Menu Name="MainMenu" DockPanel.Dock="Top">
       <MenuItem Header="File">
-        <MenuItem Header="Exit" Command="{Binding ExitCommand}" />
+        <MenuItem Header="Exit" Command="{Binding Exit}" />
```
prefer to keep unrelated changes in seperate PR.

I think this isnt releated to notifications?

pr8x replied:
It is not. I just thought that it makes it more readable.

danwalmsley replied:
I understand, but we will have to revert this change :) thanks

danwalmsley commented on samples/ControlCatalog/Pages/NotificationsPage.xaml:
```diff
@@ -3,8 +3,8 @@
-        <Button Content="Show Native Notification" Command="{Binding ShowNativeNotificationCommand}" />
+        <Button Content="Show Standard Managed Notification" Command="{Binding ShowManagedNotification}" />
+        <Button Content="Show Custom Managed Notification" Command="{Binding ShowCustomManagedNotification}" />
+        <Button Content="Show Native Notification" Command="{Binding ShowNativeNotification}" />
```
Not necessary, this code is demonstrating how to use ICommand.

pr8x committed:
!T Reverting to ReactiveCommand

pr8x committed:
Merge branch 'master' into native_notifications_win_linux_osx

pr8x committed:
!R Installing shortcut and appUserModlId in AfterSetup.

pr8x committed:
Merge branch 'native_notifications_win_linux_osx' of https://github.com/pr8x/Avalonia into native_notifications_win_linux_osx

pr8x commented on src/Avalonia.FreeDesktop/Notifications/FreeDesktopNotificationManager.cs:
```diff
@@ -0,0 +1,152 @@
+            _closeNotificationWatcher?.Dispose();
+        }
+
+        public async void Show(INotification notification)
```
`async void` is problematic for a number of reasons (Exceptions will pretty much crash the application). We should probably introduce a `ShowAsync()` function in `INotificationManager` and throw on the synchronous one. Or we adapt something like `AsyncHelper.RunSync` (https://cpratt.co/async-tips-tricks/) and just block (blocking time should be fairly short I reckon).

pr8x replied:
Same applies for `SetupWatcherTasks()` and `Connect()`. I don't really want to make the entire `X11Platform` initilization routine `async` just for these two functions.

pr8x committed:
!R Making INotificationManager.Show async

pr8x committed:
Merge branch 'master' into native_notifications_win_linux_osx

pr8x committed:
!R Using ValueTask

pr8x committed:
!T Don't create notifier every time notification is requested

kekekeks commented on src/Avalonia.FreeDesktop/Avalonia.FreeDesktop.csproj:
```diff
@@ -12,4 +12,12 @@
+  </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Tmds.DBus" Version="0.7.0" />
```
Looks like a duplicate

kekekeks commented on src/Avalonia.FreeDesktop/Notifications/FreeDesktopNotificationManager.cs:
```diff
@@ -0,0 +1,151 @@
+
+        private async ValueTask Connect()
+        {
+            _isConnected = await
```
Note that the app might be started without a session bus or without `org.freedesktop.Notifications` service on the session bus. That's the case for VNC-powered sessions, WSL, embedded devices and other scenarios without a "full" desktop environment. We should gracefully fall back to some managed implementation in that case.

Use `DBusHelper.Connection` which will be just `null` if no session DBus is available like [we are doing in our dbusmenu implementation](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.FreeDesktop/DBusMenuExporter.cs#L20). Then try to create a proxy and see if that throws. That check should be done before registering `FreeDesktopNotificationManager` in the service locator, so the rest of the code base won't be confused by dysfunctional notification manager. It would also be a good place to call `GetCapabilitiesAsync`.

pr8x replied:
> It would also be a good place to call GetCapabilitiesAsync.

Which capabilties would you check for? At first glance the features we rely on should be working in all server implementations.

pr8x replied:
Currently we're only invoking `DBusHelper.TryInitialize()` when the `UseDBusMenu` is enabled. We should probably be doing this unconditionally now.

kekekeks replied:
>Which Window should it be installed on though

The initial idea was that user code is supposed to get the notification manager from the window itself rather than using service locator. A single app might be using toplevels represented by different windowing systems. I. e. it's planned to have a "remote" mode for devtools when they are using our previewer remote widget infrastructure to render themselves to a remote machine (i. e. you will be able to connect to devtools of an application running on your android phone). So the general policy is to remove global variables when possible.

So there shouldn't be a globally accessible instance of the notification manager, only one accessed via ITopLevelImpl in some manner.

pr8x replied:
> So there shouldn't be a globally accessible instance of the notification manager, only one accessed via ITopLevelImpl in some manner.

Allright, but that seems wastly out of scope for this PR. For now the service will just return `null` when it fails to initialize. I guess the users can account for that in their code and fire up the `WindowNotificationManager` instead.

kekekeks replied:
> that seems vastly out of scope for this PR

the system notification manager should still be accessed via the TopLevel though.

pr8x replied:
Wait... So instead of registering the notification manager (on all platforms) through `AvaloniaLocator` I should just make it a property on the `TopLevel`?

kekekeks replied:
Yes, probably an attached one. The general idea is that services should be resolved via toplevels (somewhat like in Android where every damn thing needs a `Context`)

kekekeks replied:
Also, note how it works with global menu - toplevel impl implements an extra interface where it announces that it could potentially provide some service.

pr8x replied:
> toplevel impl implements an extra interface

I can't seem to find the interface you're referring to. What is it called?

kekekeks replied:
https://github.com/AvaloniaUI/Avalonia/blob/e80ddf5c6591bff9e936fa3c8655e0daff9fd39b/src/Avalonia.Controls/Platform/ITopLevelNativeMenuExporter.cs#L14

kekekeks commented on src/Avalonia.Native/NotificationManager.cs:
```diff
@@ -0,0 +1,67 @@
+            _notifications = new Dictionary<int, INotification>();
+            _notificationCloseCallback = NotificationOnClose;
+            _notificationActionCallback = NotificationOnAction;
+            _native.CloseCallback = Marshal.GetFunctionPointerForDelegate(_notificationCloseCallback);
```
Don't use function pointers with Avalonia.Native. Create a COM interface and implement it on the C# side instead. We are automatically generating native-to-managed call proxies for interfaces with `Events` suffix.

The preferred way would be to *not* use an `int` identifier field in AvnNotification but introduce
```c
AVNCOM(IAvnNotificationEvents, 21) : IUnknown
{
    virtual void OnAction() = 0;
    virtual void OnClose() = 0;
};
```
and create a new instance of a class implementing that interface for each `ShowNotification` call and pass it along with other notification info, then keep it in `ComPtr<IAvnNotificationEvents>` on the native side. That way the callback life time would be managed automatically and we won't need to keep the GC-rooted dictionary with all of the notifications.

pr8x replied:
> The preferred way would be to not use an int identifier field in AvnNotification

I guess once we implement replacing of notifications it still makes sense to have some id field to identify the OSX notification we want to replace.

> then keep it in ComPtr<IAvnNotificationEvents> on the native side

That means I have to store a vector of all the `ComPtr<IAvnNotificationEvents>` and release them when the notification closes? AFAIU you want the interface of `IAvnNotificationManager` to be `ShowNotification(AvnNotification* notification, IAvnNotificationCallbacks* callbacks)`, correct?

kekekeks replied:
+

pr8x replied:
The user can show multiple notifications. I need to store their corresponding callbacks somewhere in order to call them from the global `NotificationCenterDelegate`. I could plug them into the `NSUserNotification.userInfo` somehow, I suppose. `userInfo` expects a map of `id`, though (which is essentially a pointer to `objc_object`). So I guess I need to find some ugly workaround of putting a `ComPtr` in there...

kekekeks replied:
>userInfo expects a map of id

it seems that OSX expects the contents of userInfo to be serializable. I'm afraid that might need to change the API design to accomodate OSX notifications properly.

pr8x replied:
What API design are you thinking of? For now, we could also keep it with the `Dictionary` on C# and the function pointers. And revisit it in the future once we figured out a good/better way of designing this API.

kekekeks replied:
> For now, we could also keep it with the Dictionary on C# and the function pointers.

That can and will lead to memory leaks.

kekekeks commented on src/Windows/Avalonia.Win32/Win32Platform.cs:
```diff
@@ -19,20 +21,24 @@
+                .UseWindowingSubsystem(() => Win32.Win32Platform.Initialize(settings), "Win32")
+                .AfterSetup(b =>
+                {
+                    Win32.Win32Platform.SetupAppUserModelId(settings.AppUserModelId ?? b.Instance.Name);
```
1) this should live in `Win32.Win32Platform.Initialize`
2) add a check for the windows version, we probably don't want to do that on Win7/Win8.1

pr8x replied:
> this should live in Win32.Win32Platform.Initialize

I don't get the `Application` instance during `Win32.Win32Platform.Initialize` - it's `null`

kekekeks replied:
Then pass the AppBuilderBase instance to `Initialize` and call AfterSetup hooks there. `UseWindowingSubsystem` is called by previewer after `BuildAvaloniaApp` is already called and it overrides the windowing subsystem. So your platform-specific AfterSetup hook is called when it shouldn't be.

pr8x replied:
Do we still need the parameterless `public static void Initialize()`?

kekekeks replied:
I don't think so, the only stable public contracts of platform backends are `Use[PlatformName]` and `[PlatformName]PlatformOptions`

pr8x replied:
So should I just pass `null` as `AppBuilderBase<T>`? What is `T` supposed to be though? Also seems like a weird side-effect of calling this function that NotificationManager doesn't work.

kekekeks replied:
>So should I just pass null as AppBuilderBase<T>

You have `this T builder` right there

pr8x replied:
Oh I didn't read it correctly; I though we still need the parameterless constructor and was confused what to pass as the builder.

kekekeks reviewed (changes requested):
In addition to the comments: please, avoid doing formatting changes for the sake of doing formatted changes. It confuses `git blame` and makes it way harder to track the origin of a particular change. Same goes for renaming files. You'll probably have to squash your changes now.

kekekeks commented on native/Avalonia.Native/src/OSX/app.mm:
```diff
@@ -1,8 +1,39 @@
+- (NSString *)__bundleIdentifier;
+{
+    if (self == [NSBundle mainBundle]) {
+        return @"com.avalonia.native.osx";
```
We should add an option to configure that with `AvaloniaNativeMacOptions` and fall back to ``com.avaloniaui.apps.`+Assembly.GetEntryAssembly().Name`

kekekeks replied:
This probably should also somehow interact with bundle identifier from Info.plist if the app is running from an actual app bundle.

pr8x replied:
I guess we should check the `[[NSBundle mainBundle] bundleIdentifier]` and only install the fake bundle identifier when there isn't.

kekekeks commented:
@pr8x it seems that OSX notifications kinda "live forever" and one is expected to track them using a string `identifier` and up to 1KB of data in `userInfo`. It seems that the current API is simply not suitable for dealing with OSX notifications and might need to be redesigned completely. Not sure if that's in the scope of this PR though. 

With the current API we might want to disallow using arbitrary huge expiration times, so we could assume that notification will always expire in a reasonable time.

pr8x committed:
!T Removing duplicate package

pr8x commented:
@kekekeks  What do you mean by "live forever". They just stay in the notification center until the user manually removes them? I guess we can default to some reasonable expiration time when the user doens't specify it. But what is the problem with them just staying in the notfication center?

pr8x committed:
!R Fallback to null when FreeDesktopNotificationManager fails to initialize

kekekeks commented:
Do they stay in the notification center after application is closed?

pr8x committed:
!R Moving AfterSetup() into Win32Platform

pr8x committed:
!T Removing service check

pr8x committed:
Merge branch 'master' into native_notifications_win_linux_osx

pr8x commented:
@kekekeks Yes, the notifications indeed stay when the application exits. I should probably listen to some application shutdown event and call `removeAllDeliveredNotifications()`

kekekeks commented:
We should also call `removeAllDeliveredNotifications` on startup, I think.

kekekeks commented:
BTW, same happens on KDE.

danwalmsley commented:
closing due to inactivity... @pr8x if you would like to resolve the conflicts we can resurrect this PR.
